common/services/settings.ts

```diff
@@ -0,0 +1,113 @@
+import { EpluginSettingType, PLUGIN_SETTINGS, TpluginSetting, TPluginSettingWithKey } from '../constants';
+
+/**
+ * Get the default value of the plugin setting
+ * @param setting setting key
+ * @returns setting default value
+ */
+export function getSettingDefaultValue(setting: string) {
```

Comment:
Also, what do you think if instead of taking a string as parameter to specify the key from the PLUGIN_SETTINGS object to read, we use an enum instead:

```ts
function getDefaultValueForSetting(setting: PLUGIN_SETTINGS_KEYS)
```

This enum would contain every setting key:

```ts
enum PluginSettingsIDs {
  CRON_STATISTICS_INDEX_NAME: 'cron.statistics.index.name',
  CRON_STATISTICS_INDEX_REPLICAS: 'cron.statistics.index.replicas',
  ...,
  PATTERN: 'pattern'
}
```

then, the call to this function could be done as follows:

```ts
import { PLUGIN_SETTINGS, PluginSettingsIDs } from '../constants';

getDefaultValueForSetting(PluginSettingsIDs.PATTERN);
```

This way, we can prevent invalid values, due to typos or invalid settings (if settings are removed the linter will raise a warning).

Reply:
> Also, what do you think if instead of taking a string as parameter to specify the key from the PLUGIN_SETTINGS object to read, we use an enum instead:
> 
> ```ts
> function getDefaultValueForSetting(setting: PLUGIN_SETTINGS_KEYS)
> ```
> 
> This enum would contain every setting key:
> 
> ```ts
> enum PluginSettingsIDs {
>   CRON_STATISTICS_INDEX_NAME: 'cron.statistics.index.name',
>   CRON_STATISTICS_INDEX_REPLICAS: 'cron.statistics.index.replicas',
>   ...,
>   PATTERN: 'pattern'
> }
> ```
> 
> then, the call to this function could be done as follows:
> 
> ```ts
> import { PLUGIN_SETTINGS, PluginSettingsIDs } from '../constants';
> 
> getDefaultValueForSetting(PluginSettingsIDs.PATTERN);
> ```
> 
> This way, we can prevent invalid values, due to typos or invalid settings (if settings are removed the linter will raise a warning).

@asteriscos was commenting to me about something similar to what you are talking about. We talked about the pros and cons. He proposed to create constants that would define the keys to the configurations, and use them instead of the strings. I commented that in my opinion, these should have a similar name to their value, maybe using uppercase, as your example. In this way, it would be more readable for us.

For another hand, the current approach uses the setting key, as a string. You are right, that we could miss typing some of them. The advantage of the string, it is a little more readable than the constant. Typescript could help us to detect misstyped setting keys for this approach, but it requires work to type the variables.

We can use any proposal. Maybe we should discuss it.